src/lib.rs

```diff
@@ -8,35 +8,270 @@ mod transaction;
 pub mod update;
 
 use crate::{
-    keys::{OwnershipKeyPair, OwnershipPublicKey, PublishingKeyPair, RevocationKeyPair},
+    create::{BuildFundingPSBT, SignFundingPSBT},
+    keys::{
+        OwnershipKeyPair, OwnershipPublicKey, PublishingKeyPair, PublishingPublicKey,
+        RevocationKeyPair, RevocationPublicKey, RevocationSecretKey,
+    },
     transaction::{CommitTransaction, FundingTransaction, SplitTransaction},
 };
 use anyhow::bail;
-use bitcoin::Amount;
+use bitcoin::{Amount, Transaction};
 use ecdsa_fun::adaptor::EncryptedSignature;
-use keys::{PublishingPublicKey, RevocationPublicKey, RevocationSecretKey};
+
+#[allow(dead_code)]
+pub struct Channel {
+    x_self: OwnershipKeyPair,
+    X_other: OwnershipPublicKey,
+    pub TX_f_body: FundingTransaction,
+    pub current_state: ChannelState,
+    revoked_states: Vec<RevokedState>,
+}
+
+#[async_trait::async_trait]
+pub trait BroadcastSignedTransaction {
+    async fn broadcast_signed_transaction(&self, transaction: Transaction) -> anyhow::Result<()>;
+}
+
+#[async_trait::async_trait]
+pub trait SendMessage {
+    async fn send_message(&mut self, message: create::Message) -> anyhow::Result<()>;
+}
+
+#[async_trait::async_trait]
+pub trait ReceiveMessage {
+    async fn receive_message(&mut self) -> anyhow::Result<create::Message>;
```

Comment:
```suggestion
    async fn receive_message<M>(&mut self) -> anyhow::Result<M>;
```
An idea to avoid the matching on the `enum`: Make the message type a type parameter and just use `receive_message::<Message1>().await` on the call site.

Reply:
This is better for the call site!

Reply:
@D4nte I couldn't make this work, maybe you can give it a go once this PR is merged.

Reply:
Yes, will do, I expected it to be a pain indeed. Glad to see you did not spend too much time on it ;)

Reply:
@thomaseizinger what do you think?

I am starting to tackle that and see if we can avoid the use of the `enum Message` for the `MessageReceive` trait and get a function that returns the specific message we need instead of having to match the enum every time.

However, more I work on it, more I wonder if it's the right choice.

Reply:
I am not sure I fully see the big picture but generally, the network part of the code will usually need to be prepared receive any message at any point in time. Hence the use of an enum.

A way around that only exists if your network layer is stateful and knows, which message to deserialize next from the bytes on the wire. However, such design will usually contradict with the use of JSON or any other self-describing serialization format as it wouldn't be clear, what the source of truth is. If you want to avoid the enum, that is the way to go I think.

You can, however, also create a convenience layer that "forces" the enum into a specific message, similar to the combinators like ok_or_else on an Option turns it into a Result.

Reply:
You could probably build the receice_message trait on top of TryFrom and use that to convert the enum into its variants :)